src/apps/editor/components/matrix/form-renderer.tsx

```diff
@@ -75,31 +77,51 @@ export const FormRenderer: React.FC<Props> = ({
 
   return (
     <VCol variant="md" className={twMerge('w-full', depth > 0 && 'pl-4')}>
-      <Input
-        variant="plain"
-        defaultValue={value}
-        placeholder={type === 'array' ? 'Array key' : objectPlaceholder}
-        onBlur={onFieldUpdate(field.id, path)}
-        icon={<XIcon className="text-muted-foreground" />}
-        onIconClick={onRemoveField(field.id, path)}
-        disabled={arrayChild && type === 'object'}
-      />
-      {child.length > 0 ? (
-        <VCol variant="md" className="w-full border-l border-dashed">
-          {child.map((field) => (
-            <FormRenderer
-              key={field.id}
-              field={field}
-              path={childPath}
-              arrayChild={type === 'array'}
-              onAddField={onAddField}
-              onFieldUpdate={onFieldUpdate}
-              onRemoveField={onRemoveField}
+      <HCol className="w-full">
+        {child.length > 1 && (
+          <button onClick={() => setCollapsed(!collapsed)} className="-ml-1">
```

Comment:
This `setCollapsed(!collapsed)` is a very common mistake. React cannot ensure you that everytime that the button is clicked, the `collapsed` value will be properly updated. Supose that for every state change, would would make some calculations or React would have to trigger too many changes, which would result in a slower state update (as an example let's think 0.5s). This means that the the `collapsed` value would take at least 0.5s to be updated. Which means that, if the user clicks twice the button, in less that 0.5s of difference between clicks, the value used to update the state would not be correct, becasue it will have the old `collapsed` value 👇🏼:

```sh
# Let's see it as a sequence of events
0. Initial state of collapsed is false
1. User clicks the button
2. React receives an update call from setCollapsed with !collapsed,
   which means !false => true. As said, this update will take 0.5s.
3. User clicks again the button, with less that 0.5s of difference
4. React receives an update call from setCollapse with !collapsed,
   but since the state in the component has NOT yet changed,
   collapsed is still false, therefore !false => true. 
```

As you can see the state will not be updated accordingly and it can lead to inconsistencies in the application.

#### Solution

The signature of the dispatcher function of `useState` is:

```tsx
dispatcher<T>(T);
dispatcher<T>((prevState: T) => T);
```

The second option is a callback that will receive the previous state. In this case, we can be sure that when it is called, it will contain the correct state, always, no questions asked.

Therefore, in this case, it is recommended to use something as:

```tsx
<button onClick={() => setCollapsed((prev) => !prev)}>
 {/* */}
</button>
```

`prev` will always contain the correct state, therefore, no matter how fast the button is clicked, it will always have the proper state.

##### Additional

I have leaving functions like `onClick={() => ...}`. I prefer to have functions as handlers, defined at the top of the component, even if it is as simple as yours. I helps with readibility of the component properties 😃

Reply:
There's a section regarding this in the new React's documentation: https://react.dev/reference/react/useState#updating-state-based-on-the-previous-state